fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
+// currently writing.
+// - a *file-like* entry declares a `Size` in its `Attributes`
+//   - reads and writes are symmetric; the kernel page cache will be used, and the file size
+//     represents the current local state of the file
+//   - when writing, `data` represents the local state of the file; changes to its size will be
+//     reflected in `data` and reads will be served from it. The file's size will be `len(data)`.
+//     Unchanged sections will be filled from `plugin.Read` as needed.
+//   - when not writing, data is read directly from `plugin.Read`
+// - a *non-file-like* entry has `Size` unset
+//   - read always pulls from `plugin.Read` and writes are buffered independently
+//   - `data` stores only the data to be written, and is not initialized from `plugin.Read`
+//   - the file's size will be reported as its readable size; it will not reflect calls to `write`
+//
+// Note that writes only result in calling `plugin.Write` when a file handle is closed. Writing
+// with multiple handles will be protected by `mux`, but all writes will operate on the same `data`
+// and the first handle close will trigger `plugin.Write`.
+//
+// `readSize` will always be initialized from either the `Size` attribute, or if unset then the
+// length of data available to read.
+//
+// `writers` are used to track in-progress writes so we know when to `plugin.Write` on `Flush`
 type file struct {
-	*fuseNode
-}
+	fuseNode
 
-var _ fs.Node = (*file)(nil)
-var _ = fs.NodeOpener(&file{})
+	mux sync.Mutex
+	// Handles with in-progress writes
+	writers map[fuse.HandleID]struct{}
+	// Only valid if len(writers) > 0
+	data []byte
+	// Size of readable content, necessary for *non-file-like* entries
+	readSize uint64
+}
 
 func newFile(p *dir, e plugin.Entry) *file {
-	return &file{newFuseNode("f", p, e)}
+	return &file{fuseNode: newFuseNode("f", p, e), writers: make(map[fuse.HandleID]struct{})}
+}
+
+func (f *file) isFileLike() bool {
+	attr := plugin.Attributes(f.entry)
+	return attr.HasSize()
+}
+
+// If currently writing a file-like object, we should use local state to fulfil many requests.
+func (f *file) useLocalState() bool {
+	return len(f.writers) != 0 && f.isFileLike()
+}
+
+var _ = fs.Node(&file{})
+var _ = fs.Handle(&file{})
+
+func (f *file) Attr(ctx context.Context, a *fuse.Attr) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	if !f.useLocalState() {
+		// Fetch updated attributes only if we're not currently writing to it.
+		entry, err := f.refind(ctx)
+		if err != nil {
+			activity.Warnf(ctx, "FUSE: Attr errored %v, %v", f, err)
+			return err
+		}
+		f.entry = entry
+	}
+
+	f.fillAttr(a)
+	activity.Record(ctx, "FUSE: Attr %v: %+v", f, *a)
+	return nil
 }
 
-// Open a file for reading or writing.
+func (f *file) fillAttr(a *fuse.Attr) {
+	attr := plugin.Attributes(f.entry)
+	applyAttr(a, attr, getFileMode(f.entry))
+
+	if f.useLocalState() || !attr.HasSize() {
```

Comment:
I'd mention that `!attr.HasSize()` should only be true for non-file-like entries